src/zenml/orchestrators/step_run_utils.py

```diff
@@ -0,0 +1,466 @@
+    PipelineDeploymentResponse,
+    PipelineRunResponse,
+    PipelineRunUpdate,
+    StepRunRequest,
+    StepRunResponse,
+    StepRunUpdate,
+)
+from zenml.orchestrators import cache_utils, input_utils, utils
+from zenml.stack import Stack
+from zenml.utils import string_utils
+
+logger = get_logger(__name__)
+
+
+class StepRunRequestFactory:
+    """Helper class to create step run requests."""
+
+    def __init__(
+        self,
+        deployment: "PipelineDeploymentResponse",
+        pipeline_run: "PipelineRunResponse",
+        stack: "Stack",
+    ) -> None:
+        """Initialize the object.
+
+        Args:
+            deployment: The deployment for which to create step run requests.
+            pipeline_run: The pipeline run for which to create step run
+                requests.
+            stack: The stack on which the pipeline run is happening.
+        """
+        self.deployment = deployment
+        self.pipeline_run = pipeline_run
+        self.stack = stack
+
+    def create_request(self, invocation_id: str) -> StepRunRequest:
+        """Create a step run request.
+
+        This will only create a request with basic information and will not yet
+        compute information like the cache key and inputs. This is separated
+        into a different method `populate_request(...)` that might raise
+        exceptions while trying to compute this information.
+
+        Args:
+            invocation_id: The invocation ID for which to create the request.
+
+        Returns:
+            The step run request.
+        """
+        return StepRunRequest(
+            name=invocation_id,
+            pipeline_run_id=self.pipeline_run.id,
+            deployment=self.deployment.id,
+            status=ExecutionStatus.RUNNING,
+            start_time=datetime.utcnow(),
+            user=Client().active_user.id,
+            workspace=Client().active_workspace.id,
+        )
+
+    def populate_request(self, request: StepRunRequest) -> None:
+        """Populate a step run request with additional information.
+
+        Args:
+            request: The request to populate.
+        """
+        step = self.deployment.step_configurations[request.name]
+
+        input_artifacts, parent_step_ids = input_utils.resolve_step_inputs(
+            step=step,
+            pipeline_run=self.pipeline_run,
+        )
+        input_artifact_ids = {
+            input_name: artifact.id
+            for input_name, artifact in input_artifacts.items()
+        }
+        request.inputs = input_artifact_ids
+        request.parent_step_ids = parent_step_ids
+
+        cache_key = cache_utils.generate_cache_key(
+            step=step,
+            input_artifact_ids=input_artifact_ids,
+            artifact_store=self.stack.artifact_store,
+            workspace_id=Client().active_workspace.id,
+        )
+        request.cache_key = cache_key
+
+        # TODO: Doing this here means this will always fail when running a
```

Comment:
@htahir1

Reply:
@schustmi i think not storing the code + docstring isn't really an option as its a very nice feature people like already... any other options here?

Reply:
I can't think of anything other than what I mentioned as the alternative here, but I don't think that's an option to be honest for the sake of our request sizes. Additionally, it would be very nice to implement this feature in a way that we don't have to import the step client-side to keep our options open in the future (e.g. start a pipeline without having the requirements installed)

Reply:
I can try to hack together something that checks whether the run is started from a template, and in that case look for the docstring/source code in the steps of the run from which the template was created. This will be very hacky and also won't work anymore as soon as we allow specifying different code versions when executing a template.

Reply:
Implemented the fallback from the template run if the step importing fails.